### PR TITLE
add scalar product policy similarity protocol

### PR DESCRIPTION
**Pull request by smirnovlad**

### CODE REVIEW

File: ppotune/comm/protocols.py [hunk 1]
```diff
@@ -101,10 +102,92 @@ def update(self) -> None:
...
+        world_size = dist.get_world_size()
+        if self.self_preference is not None and world_size > 1:
+            self._weights = torch.zeros(world_size, dtype=torch.float32)
+            self._weights[dist.get_rank()] = self.self_preference
+            # Distribute remaining mass uniformly
+            remaining_mass = 1.0 - self.self_preference
+            other_weights = remaining_mass / (world_size - 1)
+            for i in range(world_size):
+                if i != dist.get_rank():
+                    self._weights[i] = other_weights
```
Comment (Arkadiy-Vladimirov) [inline ppotune/comm/protocols.py]: Для этого существуют weightage - часть протокола в которой описана развесовка. В данном случае видимо надо переиспользовать uniform_weightage

File: ppotune/comm/protocols.py [hunk 2]
```diff
@@ -101,10 +102,92 @@ def update(self) -> None:
...
+            self._weights[my_rank] = self.self_preference
+            
+            # Distribute remaining mass among other agents based on similarity
+            remaining_mass = 1.0 - self.self_preference
+            other_similarities = similarities.clone()
+            other_similarities[my_rank] = 0.0  # exclude self from softmax
+            
+            # Apply softmax to other agents' similarities
+            other_weights = torch.softmax(other_similarities / self.temperature, dim=0)
+            other_weights[my_rank] = 0.0  # ensure self weight is 0 in other_weights
```
Comment (Arkadiy-Vladimirov) [inline ppotune/comm/protocols.py]: Точно мы хотим замешивать софтмакс сюда?) Лучше наверное сначала простой вариант без него. Ну и опять же для этого есть softmax weightage

File: ppotune/comm/protocols.py [hunk 3]
```diff
@@ -101,10 +102,92 @@ def update(self) -> None:
...
+        
+        if self.self_preference is not None and world_size > 1:
+            # Set fixed self preference
+            self._weights = torch.zeros(world_size, dtype=torch.float32, device='cuda')
+            self._weights[my_rank] = self.self_preference
+            
+            # Distribute remaining mass among other agents based on similarity
+            remaining_mass = 1.0 - self.self_preference
+            other_similarities = similarities.clone()
+            other_similarities[my_rank] = 0.0  # exclude self from softmax
```
Comment (Arkadiy-Vladimirov) [inline ppotune/comm/protocols.py]: Много боилерплейта которого не должно быть при правильном использовании weightage

File: ppotune/comm/protocols.py [hunk 4]
```diff
@@ -101,10 +102,92 @@ def update(self) -> None:
...
+        # Compute similarities with all agents (including self)
+        similarities = torch.zeros(world_size, dtype=torch.float32, device='cuda')
+        for peer in range(world_size):
+            sim = 0.0
+            for k in range(len(self._policy_batch)):
+                a = peer_policies[my_rank][k].to('cuda')
+                b = peer_policies[peer][k].to('cuda')
+                min_len = min(len(a), len(b))
+                dot = torch.dot(a[:min_len], b[:min_len]).item()
+                sim += dot
```
Comment (Arkadiy-Vladimirov) [inline ppotune/comm/protocols.py]: Лучше без фора)

File: ppotune/comm/protocols.py [hunk 5]
```diff
@@ -101,10 +102,92 @@ def update(self) -> None:
...
+        self._policy_batch = policy_batch
+
+    def gather(self, stats):
+        pass
+
+    def update(self):
+        if self._policy_batch is None:
+            raise ValueError("Policy batch not set")
+        peer_policies = [None] * dist.get_world_size()
+        dist.all_gather_object(peer_policies, self._policy_batch)
```
Comment (Arkadiy-Vladimirov) [inline ppotune/comm/protocols.py]: Плохой нейминг. Это не peer policies, это референсные батчи с них

File: ppotune/comm/protocols.py [hunk 6]
```diff
@@ -101,10 +102,92 @@ def update(self) -> None:
...
+
+    def __call__(self, tensors):
+        tensors = torch.stack(tensors)
+        weights = self._weights.to(tensors[0].device)
+        log.collect("self_preference", weights[dist.get_rank()])
+
+        for _ in range(tensors.dim() - 1):
+            weights = weights.unsqueeze(-1)
+        return (weights * tensors).sum(dim=0)
+
```
Comment (Arkadiy-Vladimirov) [inline ppotune/comm/protocols.py]: У базового класса есть уже есть кол метод. Скорее всего если все правильно написать то перегружать его не придется

File: ppotune/data/loaders.py [hunk 7]
```diff
@@ -121,3 +121,47 @@ def distributed_dataloader(
...
+        tokens_key="tokens",
+        pad_token=tokenizer.pad_id
+    )
+    return DataLoader(
+        dataset=dataset,
+        sampler=sampler,
+        batch_size=batch_size,
+        drop_last=True,
+        collate_fn=collator
+    )
```
Comment (Arkadiy-Vladimirov) [inline ppotune/data/loaders.py:158]: Очень не хочется заводить под это отдельный даталодер лучше подаправить существующий чтобы в нем можно было раздать всем одинаковый сид

File: ppotune/comm/protocols.py [hunk 8]
```diff
@@ -101,10 +102,92 @@ def update(self) -> None:
...
+            other_weights = remaining_mass / (world_size - 1)
+            for i in range(world_size):
+                if i != dist.get_rank():
+                    self._weights[i] = other_weights
+        else:
+            # Default: uniform weights
+            self._weights = torch.ones(world_size) / world_size
+
+    def set_policy_batch(self, policy_batch):
+        self._policy_batch = policy_batch
```
Comment (Arkadiy-Vladimirov) [inline ppotune/comm/protocols.py]: Для этого можно нужно использовать метод gather. Нужно чтобы у всех policy mixture был один интерфейс

File: receipts/ppo/ppo.py [hunk 9]
```diff
@@ -367,6 +407,15 @@ def train(self) -> None:
             self.eval(self.policy, step)
 
             self._ref_policy.gather(trajectory)
+            if hasattr(self._ref_policy, 'set_policy_batch_for_similarity'):
```
Comment (Arkadiy-Vladimirov) [inline receipts/ppo/ppo.py]: Такого быть не должно, у всех реф полиси должен быть один интерфейс. Все статистики можно собирать через gather см. коммент к similarity mixture классу

File: receipts/ppo/ppo.py [hunk 10]
```diff
@@ -367,6 +407,15 @@ def train(self) -> None:
...
+                try:
+                    test_batch = next(self.similarity_iter)
+                except StopIteration:
+                    self.similarity_iter = iter(self.similarity_dataloader)
+                    test_batch = next(self.similarity_iter)
```
Comment (Arkadiy-Vladimirov) [inline receipts/ppo/ppo.py]: не очень понятно зачем это надо. из similarity даталодера должен получаться неистощаемый итератор

File: receipts/ppo/ppo.py [hunk 11]
```diff
@@ -367,6 +407,15 @@ def train(self) -> None:
...
+                policy_batch = self.generate_policy_batch(test_batch)
+                self._ref_policy.set_policy_batch_for_similarity(policy_batch)
+                torch.cuda.empty_cache()
```
Comment (Arkadiy-Vladimirov) [inline receipts/ppo/ppo.py]: все это должно происходить внутри similarity policy mixture чтобы не нарушать общий api и не загромождать основной скрипт

File: receipts/ppo/ppo.py [hunk 12]
```diff
@@ -318,6 +329,35 @@ def generate_trajectory_batched(
...
+            response_mask = generated.response_mask[:, query_len:]
+            
+            for b in range(logprobs.shape[0]):
+                length = response_mask[b].sum().item()
+                policy_batch.append(logprobs[b, :length])
+                
+            torch.cuda.empty_cache() if self._empty_cache else None
+        
+        return policy_batch
+
```
Comment (Arkadiy-Vladimirov) [inline receipts/ppo/ppo.py]: наверное должно работать, но выглядит громоздко и должно происходить внутри similarity policy mixture. возможно можно как то переиспользовать логику генерации из скрипта выше. по крайней мере выделить в отдельную функцию

File: receipts/ppo/ppo.py (hunk 11 above)
Comment (Arkadiy-Vladimirov) [inline receipts/ppo/ppo.py]: ВАЖНО! нет, мы хотели не так сделать. так получается что у всех разные батчи токенов в зависимости от того кто что нагенерил. Да и вообще генерация это довольно долго не хочется чтобы это здесь происходило тоже. Мы хотели просто таскать бачти с референсыми вопросами И ОТВЕТАМИ и смотреть какие на них логиты у моделей.

**Review by Arkadiy-Vladimirov — CHANGES_REQUESTED** [review]
Нужно поправить логику, это не совсем так как мы хотели сделать. Код тоже лучше порефакторить, сейчас сложновато читать я мог что то упустить из багов
